### PR TITLE
[DEV-2435] enable observation table to be associated with multiple use cases

### PR DESCRIPTION
**Pull request by yanhuisg**
## Type of Change

Label this pull request to place it under the correct category in Release Notes:

|        **Pull Request Label**         | **Category in Release Notes** |
|:-------------------------------------:|:-----------------------------:|
|       `enhancement`, `feature`        |          🚀 Features          |
| `bug`, `refactoring`, `bugfix`, `fix` |    🔧 Fixes & Refactoring     |
|       `build`, `ci`, `testing`        |    📦 Build System & CI/CD    |
|              `breaking`               |      💥 Breaking Changes      |
|            `documentation`            |       📝 Documentation        |
|            `dependencies`             |    ⬆️ Dependencies updates    |

## Checklist

- [ ] I have read the [`CODE_OF_CONDUCT.md`](https://github.com/featurebyte/featurebyte/blob/main/CODE_OF_CONDUCT.md) and [`CONTRIBUTING.md`](https://github.com/featurebyte/featurebyte/blob/main/CONTRIBUTING.md) guides.
- [ ] I have written tests for the changes made.
- [ ] I have written docstrings in [NumpyDoc format](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard)
- [ ] I have labeled my Pull Request correctly

### CODE REVIEW

File: featurebyte/api/use_case.py [hunk 1]
```diff
@@ -149,8 +149,8 @@ def add_observation_table(self, observation_table_name: str) -> None:
         >>> use_case.add_observation_table(observation_table_name)  # doctest: +SKIP
         """
         observation_table = ObservationTable.get(observation_table_name)
-        self.update(
-            update_payload={"new_observation_table_id": observation_table.id},
+        observation_table.update(
+            update_payload={"use_case_id_to_add": self.id, "use_case_ids": None},
```
Comment (gee-senbong) [inline featurebyte/api/use_case.py:153]: 👍

File: featurebyte/service/observation_table.py [hunk 2]
```diff
@@ -359,27 +363,77 @@ async def update_observation_table(
...
+                    )
+
+                use_case_ids.append(data.use_case_id_to_add)
+
+            # validate use case id to remove
+            if data.use_case_id_to_remove:
+                if data.use_case_id_to_remove not in use_case_ids:
+                    raise ObservationTableInvalidUseCaseError(
+                        f"Cannot remove UseCase {data.use_case_id_to_remove} as it is not associated with the ObservationTable."
+                    )
```
Comment (gee-senbong) [inline featurebyte/service/observation_table.py:416]: Would be good to have an unit test to cover this.

File: featurebyte/service/observation_table.py [hunk 3]
```diff
@@ -359,27 +363,77 @@ async def update_observation_table(
...
 
+        if data.use_case_ids:
+            raise ObservationTableInvalidUseCaseError("use_case_ids is not a valid field to update")
+
+        observation_table = await self.get_document(document_id=observation_table_id)
+        if data.use_case_id_to_add or data.use_case_id_to_remove:
+            if not observation_table.context_id:
+                raise ObservationTableInvalidUseCaseError(
+                    f"Cannot add/remove UseCase as the ObservationTable {observation_table_id} is not associated with any Context."
+                )
```
Comment (gee-senbong) [inline featurebyte/service/observation_table.py:378]: Would be good to have an unit test to cover this.

File: featurebyte/service/observation_table.py (hunk 2 above)
Comment (yanhuisg) [reply featurebyte/service/observation_table.py:416]: I think the test is already in this PR: https://github.com/featurebyte/featurebyte/blob/yanhui/DEV-2435_2/tests/unit/routes/test_observation_table.py#L244

File: featurebyte/service/observation_table.py (hunk 3 above)
Comment (yanhuisg) [reply featurebyte/service/observation_table.py:378]: The test is here: https://github.com/featurebyte/featurebyte/blob/yanhui/DEV-2435_2/tests/unit/routes/test_observation_table.py#L263

**Review by gee-senbong — APPROVED** [review]

**Pull request merged**